lib/mirage_block_partition.ml

```diff
@@ -70,19 +72,31 @@ module Make(B : Mirage_block.S) = struct
     let sector_mid = Int64.add sector_start first_sectors in
     if sector_mid > sector_end then
       raise (Invalid_argument "Partition point beyond device");
-    ({ b; sector_size; sector_start; sector_end = sector_mid },
-        { b; sector_size; sector_start = sector_mid; sector_end })
+    ({ b; sector_size; sector_start; sector_end = sector_mid; connected = true },
+        { b; sector_size; sector_start = sector_mid; sector_end; connected = true })
 
   let connect first_sectors b =
     let open Lwt.Syntax in
     let+ { Mirage_block.sector_size; size_sectors = sector_end; _ } = B.get_info b in
     let sector_start = 0L in
     partition b ~sector_size ~sector_start ~sector_end ~first_sectors
 
-  let subpartition first_sectors { b; sector_size; sector_start; sector_end } =
-    partition b ~sector_size ~sector_start ~sector_end ~first_sectors
+  (* let subpartition first_sectors { b; sector_size; sector_start; sector_end; connected = true } =
+    partition b ~sector_size ~sector_start ~sector_end ~first_sectors *)
+    let subpartition first_sectors { b; sector_size; sector_start; sector_end; connected } =
+  match connected with
+  | true ->
+      partition b ~sector_size ~sector_start ~sector_end ~first_sectors
+  | false ->
+      
+      failwith "unconnected block"
 
-  let disconnect b =
-    (* XXX disconnect both?! *)
-    B.disconnect b.b
+let disconnect b =
+  if b.connected then (
+    b.connected <- false;
+    Lwt.return_unit
+  )
+  else (
+    Lwt.fail_with "Partition is disconnected"
+  )
```

Comment:
Good! I think it should be OK to call disconnect on an already disconnected device - so we shouldn't need to raise an exception.

The documentation from `Mirage_block.S.disconnect`:

    Disconnect from the device. While this might take some time to              
    complete, it can never result in an error.